### PR TITLE
Add support for Custom Accelerate Instance in Trainer

### PR DESCRIPTION
**Pull request by Flakes342**
# What does this PR do?

Fixes #40616 

This PR adds the ability to pass a **custom Accelerator instance** to the Trainer constructor, giving users fine-grained control over distributed training configurations while maintaining all existing Trainer functionality.
## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #40616 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Description
Currently, users who need advanced distributed training configurations (custom DeepSpeed settings, specific FSDP configurations, etc.) have limited options to customize the _Accelerator_ used by Trainer. The existing approach forces users to either modify TrainingArguments which doesn't expose all accelerate options to them or use accelerator_config which is less intuitive and limited.

This change allows users to create and configure their own Accelerator instance with full control over:
- DeepSpeed configurations with custom plugins
- FSDP settings and policies
- Tensor parallelism configurations
- Custom dataloader behaviors
- Advanced gradient accumulation strategies
- Experiment tracking and logging setup

## Changes
- Modified `Trainer` initialization to add an optional `accelerator` parameter
- Added `postprocess_accelerator` method: Handles postprocessing for user-provided accelerators - similar to current `create_accelerator_and_postprocess` method
- Refactored accelerator initialization: Clean separation between creation and postprocessing logic

## Example

Adding a simple example of how custom accelerator functionality can be used in trainer:
```
from accelerate import Accelerator
from accelerate.utils import DeepSpeedPlugin

# Custom DeepSpeed configuration
deepspeed_plugin = DeepSpeedPlugin(
    zero_stage=3,
    offload_optimizer_device="cpu",
    offload_param_device="cpu",
    zero3_init_flag=True,
    zero3_save_16bit_model=True
)

accelerator = Accelerator(deepspeed_plugin=deepspeed_plugin)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=dataset,
    accelerator=accelerator
)
```
## Testing
Added comprehensive tests to _trainer/test_trainer.py_ covering:
- Basic custom accelerator functionality
- Backward compatibility (no accelerator parameter)
- Postprocessing attribute setup
- Settings preservation
- Gather function configuration
- End-to-end training integration

Please let me know if the tests need to be written in some other way. Also, please review and provide your comments and I'd be happy to incorporate any corrections/suggestions.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -5539,6 +5547,71 @@ def create_accelerator_and_postprocess(self):
         ):
             raise ValueError("save_only_model option is not compatible with FSDP state dict type 'SHARDED_STATE_DICT'")
 
+
+    def postprocess_accelerator(self):
```
Comment (Flakes342) [inline src/transformers/trainer.py:5566]: I wanted to create this as a different method to support any modifications to custom accelerate function in future too.

**Comment by Flakes342** [comment]
Again, I am not sure what these test failures are..

**Comment by Flakes342** [comment]
Changes for this PR have been done in:
src/transformers/trainer.py
tests/trainer/tests_trainer.py

Rest all seems to have happened because of auto formatting fix

**Comment by qgallouedec** [comment]
Thanks for the contribution, I'm closing, see https://github.com/huggingface/transformers/issues/40616#issuecomment-5122831361 and https://github.com/huggingface/transformers/issues/40616#issuecomment-3329067789
